configs/module.go

```diff
@@ -491,3 +491,16 @@ func (m *Module) ProviderForLocalConfig(pc addrs.LocalProviderConfig) addrs.Prov
 	}
 	return addrs.NewLegacyProvider(pc.LocalName)
 }
+
+func (m *Module) ProviderForResource(r *Resource) addrs.LocalProviderConfig {
+	if r.ProviderConfigRef == nil {
+		return addrs.LocalProviderConfig{
+			LocalName: m.LocalNameForProvider(r.Addr().DefaultProvider()),
```

Comment:
The general idea of this function seems good to me. I'm a little unsure about this particular part of it, though... it feels a little funny to me because right now `r.Addr().DefaultProvider()` is returning a legacy provider FQN, which allows this to work in our current not-quite-there state, but the evolution of this to a world with proper FQNs feels a little unclear to me.

I guess the crux of what is bothering me is that we no longer really have enough information in a resource type like `aws_instance` or `happycloud_vm` to reliably select a "default FQN" for it.

One possibility would be to assume that the prefix always identifies a _default_ (in the `hashicorp` namespace) provider, so that the above would default to `hashicorp/aws` and `hashicorp/happycloud` respectively. But that would imply that for any provider _outside_ the `hashicorp` namespace we would now require an explicit `provider` argument in each of the resources, like this:

```hcl
terraform {
  required_providers = {
    happycloud = {
      source = "happycorp/happycloud"
    }
  }
}

resource "happycloud_vm" "example" {
  # This is mandatory because otherwise we'd select hashicorp/happycloud,
  # which either doesn't exist or is some other provider entirely.
  provider = happycloud
}
```

Off the top of my head right now I'm honestly not really sure what to do about this in the long run. Treating the prefix of the resource type name as a _local name_ doesn't feel right either, because the prefix of the resource type name is something chosen within the provider binary itself, not something user-customizable:

```hcl
terraform {
  required_providers = {
    myhappycloud = {
      source = "happycorp/happycloud"
    }
  }
}

resource "happycloud_vm" "example" {
  # Because the local name doesn't match the prefix hard-coded into
  # the provider, now the user has to explicitly set it.
  provider = myhappycloud
}
```

That at least feels easier to _explain_, even if it's still a little conceptually odd. I'd explain it like: most of the time you should make sure the local name matches the provider type name, in which case you can omit `provider` arguments as long as the resource types match the provider type name.

I see one final option, which is _also_ rather odd but maybe most consistent with how we've set up the rest of this model: find the provider FQN for `happycloud_vm` by filtering all of the FQNs used by this module (explicit or implied) for ones whose `Type` is set to `happycloud`. If the resulting set contains only one item, select that FQN. Otherwise, produce an error requiring the user to disambiguate with `provider` arguments:

```hcl
terraform {
  required_providers = {
    myhappycloud = {
      source = "happycorp/happycloud"
    }
  }
}

resource "happycloud_vm" "example" {
  # Don't need to set "provider" here because there's exactly one
  # explicitly-declared FQN whose type is "happycloud" above.
}
```

```hcl
terraform {
  required_providers = {
    # (nothing)
  }
}

resource "aws_instance" "example" {
  # Don't need to set "provider" here because there are no
  # explicitly-declared FQNs whose Type are "aws", and so
  # we can assume the user means "hashicorp/aws".
}
```

```hcl
terraform {
  required_providers = {
    myaws = {
      source = "awesomecorp/aws"
    }
  }
}

resource "aws_instance" "example" {
  # Don't need to set "provider" here because there is exactly
  # one explicitly-declared FQN whose type is "aws".
  # We would select awesomecorp/aws in this case.
}
```

```hcl
terraform {
  required_providers = {
    happycloud = {
      source = "happycorp/happycloud"
    }
    myhappycloud = {
      source = "awesomecorp/happycloud"
    }
  }
}

resource "happycloud_vm" "example" {
  # We _do_ need to set "provider" for this one, because there
  # are two explicitly-declared FQNs whose Type are "happycloud"
  provider = happycloud
}
```

Although this final option does seem like the one that is most consistent with the model as we've defined it, it also seems like the one that is the most surprising and hard to explain.

I find myself currently leaning towards the second option, where we'd treat the prefix like a _local name_ and require the user to set `provider` if they've selected an unconventional local name for the provider, because it seems like that one does the best job of allowing omission of the `provider` argument in "normal" use-cases where the Type part of all of the FQNs in use are unique and match the local names. I'm not totally convinced though, and would love to talk about this some more if you didn't reach the same conclusion! :thinking:

Reply:
I think I follow what you're saying, but I want to clarify things. The following examples are based on the various design documents that precluded this work, and as always it *can* be changed, but this is the plan up until now.

Any time the provider's local name does not match the implied name, the `provider` attribute is required:

```
terraform {
  required_providers = {
    myaws = {
      source = "awesomecorp/aws"
    }
  }
}

resource "aws_instance" "example" {
  provider = myaws
}
``` 

`required_providers`  is required if the provider is not a default hashicorp provider:
```
terraform {
  required_providers = {
    aws = {
      source = "awesomecorp/aws"
    }
  }
}

resource "aws_instance" "example" {
  // no provider attribute required
}
``` 

We will not allow any ambiguity or magic mapping of local name - type - default type, etc.
Based on that, we have the following scenarios:

* a resource with no `provider` attribute: terraform will compare the implied provider name (the `aws` in `aws_instance` with the information gathered from `required_providers`. If nothing is found, we can assume that aws = hashicorp/aws. 

* a resource has a `provider` attribute, so terraform can use the ProviderForLocalConfig function to get the provider fqn (in this case, looking up the provider from the local name directly)


I *think* that this explains the missing information you mentioned in your comment. We _can_ assume a default hashicorp provider when: there is no provider config in the resource, nor a match by the implied provider type in `required_providers`.

Reply:
Cool... it sounds like you're describing the second variant I was thinking about in my comment, and the one I was thinking seemed the least objectionable, which makes me feel better about that answer.

Reply:
oh, that's a great suggestion, thank you. I definitely favor returning a string to make it _extra clear_ that we are returning a best guess - that's exactly the kind of change that will make this more intuitive (ok, less unintuitive) for developers coming across this the first time.